pkgs/applications/version-management/sapling/gen-deps.py

```diff
@@ -1,18 +1,47 @@
 #!/usr/bin/env nix-shell
-#!nix-shell -i python3 -p "python3.withPackages (ps: with ps; [ requests ])"
+#!nix-shell -i python3 -p cargo -p "python3.withPackages (ps: with ps; [ requests ])"
 import json
+import pathlib
 import re
+import tempfile
+import os
+import shutil
 from hashlib import sha1
 from struct import unpack
 from subprocess import run
+import subprocess
 
 from requests import get
 
 # Fetch the latest stable release metadata from GitHub
-latestTag = get("https://api.github.com/repos/facebook/sapling/releases/latest").json()[
-    "tag_name"
-]
+releaseMetadata = get("https://api.github.com/repos/facebook/sapling/releases/latest").json()
+latestTag = releaseMetadata["tag_name"]
+latestTarballURL = releaseMetadata["tarball_url"]
 
+[_tarballHash, sourceDirectory] = run(
+    ["nix-prefetch-url", "--print-path", "--unpack", latestTarballURL],
+    check=True,
+    text=True,
+    stdout=subprocess.PIPE,
+).stdout.rstrip().splitlines()
+
+def updateCargoLock():
+    with tempfile.TemporaryDirectory() as tempDir:
+        tempDir = pathlib.Path(tempDir)
+
+        # NOTE(strager): We cannot use shutil.tree because it copies the
+        # read-only permissions.
```

Comment:
Why are we not using `shutil.copytree` followed by a `chmod` on the directory? I have had success with this:
```python
def updateCargoLock():
    with tempfile.TemporaryDirectory() as tempDir:
        tempDir = pathlib.Path(tempDir)
        shutil.copytree(sourceDirectory, tempDir, dirs_exist_ok=True)
        os.chmod(tempDir / "eden" / "scm", 0o755)
        run(["cargo", "fetch"], check=True, cwd=tempDir / "eden" / "scm")
        shutil.copy(tempDir / "eden" / "scm" / "Cargo.lock", "Cargo.lock")
```

Reply:
That solution felt nasty to me. Copying with the *correct* permissions seemed better than copying with the *wrong* permissions then fixing them.